### PR TITLE
AST: Buff Extension module. 

### PR DESCRIPTION
**Pull request by suShirow**
It's looking halfway decent now, going to park this here for implementation review (sketchy) and for other devs to input while I work on other modules.
Major concern is how to wait for the delayed aoe effect of CO to resolve on every target.
Minor concern is having to write another set of checks for buff refreshes on pets.

Added getStatuses() to Entity.js

### CODE REVIEW

File: src/parser/core/Entity.js [hunk 1]
```diff
@@ -38,4 +38,27 @@ export default class Entity {
...
+
+
+		const currentTimestamp = forTimestamp < this.parser.currentTimestamp ? forTimestamp
+			: this.parser.currentTimestamp
+
+		if (forTimestamp < this.parser.currentTimestamp) {
+			console.warn('Entity:getStatuses timestamp given is later than current timestamp')
+		}
+
+		return this.buffs.filter(buff =>
```
Comment (ackwell) [inline src/parser/core/Entity.js]: Might be worth consolidating this last bit of logic w/ the `getStatus` method in some way?

File: src/parser/jobs/ast/About.js [hunk 2]
```diff
@@ -1,18 +1,33 @@
...
+import ACTIONS from 'data/ACTIONS'
+
+import CONTRIBUTORS from 'data/CONTRIBUTORS'
+
 import CoreAbout from 'parser/core/modules/About'
 
 export default class About extends CoreAbout {
 	description = <Fragment>
-		<p>This analyser aims to identify some of the low-hanging fruit that could be used to improve your gameplay, as well as (eventually) (hopefully) give deeper insight into optimizing your healing and card usage.</p>
+		<p>The biggest <ActionLink {...ACTIONS.DRAW} /> to an AST is their ability to buff their party DPS with Arcanum.
```
Comment (ackwell) [inline src/parser/jobs/ast/About.js]: I should have you arrested for this.

File: src/parser/jobs/ast/BuffExtensions.module.css [hunk 3]
```diff
@@ -0,0 +1,24 @@
+.buffextensions__headerItem {
```
Comment (ackwell) [inline src/parser/jobs/ast/BuffExtensions.module.css]: BEM :blobknife:

No need to namespace css classes, the names get mangled to include module info and a hash during build.

File: src/parser/jobs/ast/BuffExtensions.js [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+	 * @return {void} null
+	 */
+	_onBuffRefresh(event) {
+		const statusID = event.ability.guid
+
+		// Ignore if timestamp is after aoe effect grace period
+		if (this._oppositionEvent && event.timestamp < (this._oppositionEvent.event.timestamp + this._envEffectGracePeriod)) {
+
+			// Ignore refreshes on protects, royal road statuses,
+			if (statusID !== STATUSES.PROTECT.id
```
Comment (ackwell) [inline src/parser/jobs/ast/BuffExtensions.js]: You can consolidate this up top of the file with something like

```js
const IGNORE_STATUSES = [
	STATUSES.PROTECT.id,
	STATUSES.COLLECTIVE_UNCONSCIOUS.id,
	...ROYAL_ROAD_STATES,
	...HELD_ARCANA,
	...DRAWN_ARCANA,
]
```

That way you only need to do the one `IGNORE_STATUSES.includes(statusId)` down here.

File: src/parser/jobs/ast/BuffExtensions.js [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+	 * Populates the _oppositionEvent created in _onOpposition with refresh events on friendlies
+	 *
+	 * @param {obj} log event of a Celestial Opposition cast
+	 * @return {void} null
+	 */
+	_onBuffRefresh(event) {
+		const statusID = event.ability.guid
+
+		// Ignore if timestamp is after aoe effect grace period
+		if (this._oppositionEvent && event.timestamp < (this._oppositionEvent.event.timestamp + this._envEffectGracePeriod)) {
```
Comment (ackwell) [inline src/parser/jobs/ast/BuffExtensions.js]: You could flatten this function a lot by reversing the `if`s and shortcutting out, i.e.:

```js
if (cond) {
	// ...
}

// vs

if (!cond) { return }
// ...
```

File: src/parser/jobs/ast/BuffExtensions.js [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+				})
+			}
+		}
+
+
+
+	}
+
+	output() {
+		const panels = this._dilationUses.map(dilation => {
```
Comment (ackwell) [inline src/parser/jobs/ast/BuffExtensions.js]: I'd try to flatten this a little. A lot of logic looping around here.

File: src/parser/jobs/ast/BuffExtensions.js (hunk 5 above)
Comment (suShirow) [reply src/parser/jobs/ast/BuffExtensions.js]: I really needed this. Always wondered which was the cleaner way to write these

**Comment by suShirow** [comment]
I've updated `AoE.js` to make aoe events for `heal, applybuff, refreshbuff` events. I learned a lot about how AoEs are presented in the log events.

Problem: Then I tried to use `aoerefreshbuff` for my Celestial opposition. I found out that buff events have a longer pulse gap than damage/heal events. I also realized that it returned separate events for each status, which is actually working as intended. It got pretty sticky when I tried to merge them into a single `OnCelestialOpposition` event, because each event now has the pulse thresholds and now we have to worry about correlating them all over again.

Conclusion: The logic is simpler if I just identified the players affected by CO through the `refreshbuff` events, and then checked their statuses for what they had from the Astro in the moments around the CO cast. 
This is accurate to a reasonable degree, the module works, and we can come up with some suggestions derived from the information pulled. 
I just hope the `AoE.js` is working well.

I've spent more than a couple of days on this and would like to move onto Minor Arcanas usage. 
We can come back to this when the implementation causes seeing organs to bleed spontaneously.

**Pull request merged**